secure/storage/src/storage.rs

```diff
@@ -47,33 +50,137 @@ pub trait Storage: Send + Sync {
+            Value::Ed25519PrivateKey(private_key) => {
+                let (new_private_key, new_public_key) = new_ed25519_key_pair()?;
+                self.set(
+                    &get_previous_version_name(key_pair_name),
+                    Value::Ed25519PrivateKey(private_key),
+                )?;
+                self.set(key_pair_name, Value::Ed25519PrivateKey(new_private_key))?;
+                Ok(new_public_key)
+            }
             _ => Err(Error::UnexpectedValueType),
         }
     }
 
+    /// Signs the given message using the private key associated with the given 'key_pair_name'.
+    /// If the key pair doesn't exist, or the caller doesn't have the appropriate
+    /// permissions to retrieve and use the public key, this call will fail with an error.
+    fn sign_message(
+        &mut self,
+        key_pair_name: &str,
+        message: &HashValue,
+    ) -> Result<Ed25519Signature, Error> {
+        let private_key = self.get_private_key_for_name(key_pair_name)?;
+        Ok(private_key.sign_message(message))
+    }
+
+    /// Signs the given message using the private key associated with the given 'key_pair_name'
+    /// and 'key_pair_version'. If the key pair doesn't exist, or the caller doesn't have the
+    /// appropriate permissions to perform the operation, this call will fail with an error.
+    /// Note: the 'key_pair_version' is specified using the public key associated with a key pair.
+    /// Only two versions of a key pair are ever maintained at any given time (i.e., the
+    /// current version, and the previous version).
+    fn sign_message_using_version(
+        &mut self,
+        key_pair_name: &str,
+        key_pair_version: Ed25519PublicKey,
+        message: &HashValue,
+    ) -> Result<Ed25519Signature, Error> {
+        let private_key =
+            self.get_private_key_for_name_and_version(key_pair_name, key_pair_version)?;
+        Ok(private_key.sign_message(message))
+    }
+
     /// Resets and clears all data held in the storage engine.
     /// Note: this should only be exposed and used for testing. Resetting the storage engine is not
     /// something that should be supported in production.
     #[cfg(test)]
     fn reset_and_clear(&mut self) -> Result<(), Error>;
 }
+
+/// Helper method to generate a new ed25519 key pair using entropy from the OS.
+fn new_ed25519_key_pair() -> Result<(Ed25519PrivateKey, Ed25519PublicKey), Error> {
+    let mut seed_rng = OsRng::new().map_err(|e| Error::EntropyError(e.to_string()))?;
+    let mut rng = rand::rngs::StdRng::from_seed(seed_rng.gen());
+    let private_key = Ed25519PrivateKey::generate_for_testing(&mut rng);
+    let public_key = private_key.public_key();
+    Ok((private_key, public_key))
+}
+
+/// Helper method to get the name of the previous version of the given key pair, as held in
+/// secure storage.
+fn get_previous_version_name(key_pair_name: &str) -> String {
```

Comment:
is this something we want at this layer or the layer below? I had imagined we would have passed the public key information into the inner implementation.

Reply:
I think for now it makes sense for this to live here (as it's common across all three backends). We may want to push the public key information further down in the future (e.g., for vault transit), but I think this is something we only need to consider on a backend-by-backend basis.

Reply:
Let me resolve these :P

I don't really think that makes sense. Versions are inherent to the backend. Is there a reason we aren't leveraging the versions within transit?